src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_StringInterpolation.cs

```diff
@@ -77,48 +112,110 @@ private void MakeInterpolatedStringFormat(BoundInterpolatedString node, out Boun
-            BoundExpression format;
-            ArrayBuilder<BoundExpression> expressions;
-            MakeInterpolatedStringFormat(node, out format, out expressions);
-            if (expressions.Count == 0)
+
+            BoundExpression result;
+
+            if (CanLowerToStringConcatenation(node))
             {
-                // There are no fill-ins. Handle the escaping of {{ and }} and return the value.
-                Debug.Assert(!format.HasErrors && format.ConstantValue != null && format.ConstantValue.IsString);
-                var builder = PooledStringBuilder.GetInstance();
-                var formatText = format.ConstantValue.StringValue;
-                int formatLength = formatText.Length;
-                for (int i = 0; i < formatLength; i++)
+                // All fill-ins, if any, are strings, and none of them have alignment or format specifiers.
+                // We can lower to a more efficient string concatenation
+                // The normal pattern for lowering is to lower subtrees before the enclosing tree. However in this case
+                // we want to lower the entire concatenation so we get the optimizations done by that lowering (e.g. constant folding).
+
+                int length = node.Parts.Length;
+                if (length == 0)
                 {
-                    char c = formatText[i];
-                    builder.Builder.Append(c);
-                    if ((c == '{' || c == '}') && (i + 1) < formatLength && formatText[i + 1] == c)
+                    // $"" -> ""
+                    return _factory.StringLiteral("");
+                }
+
+                if (length == 1)
+                {
+                    var part = node.Parts[0];
+
+                    if (part is BoundStringInsert fillin)
+                    {
+                        // $"{part}"
+                        // this is the filled-in expression
+                        part = fillin.Value;
+
+                        if (part.ConstantValue != null)
+                        {
+                            if (part.ConstantValue.StringValue == null)
+                            {
+                                return _factory.StringLiteral("");
+                            }
+                            else
+                            {
+                                result = part;
+                            }
+                        }
+                        else if (part is BoundBinaryOperator bbo && bbo.OperatorKind == BinaryOperatorKind.StringConcatenation)
+                        {
+                            // $"{a + b}" -> a + b
+                            result = part;
+                        }
+                        else
+                        {
+                            result = _factory.Coalesce(part, _factory.StringLiteral(""));
+                        }
+                    }
+                    else
+                    {
```

Comment:
>{ [](start = 20, length = 1)

Consider asserting that the part is a `BoundLiteral` and ConstantValue is not null (ie. a string literal)

Reply:
I agree... please do that.

---
In reply to: [190382117](https://github.com/dotnet/roslyn/pull/26612#discussion_r190382117) [](ancestors = 190382117)

Reply:
The existing lowering to `string.Format` doesn't have that assert. Should it be added there too?

Reply:
Yes, that would be good. Thanks

---
In reply to: [190494041](https://github.com/dotnet/roslyn/pull/26612#discussion_r190494041) [](ancestors = 190494041)